[READY] Email System

atlantiscze opened the pull request:
- This system is, eventually, suposed to replace PDA messages. It is slightly more complicated to use than "Click'n Send" PDAs, but compensates for that with other benefits.
- Each player is given an email account and password upon joining the game (john.smith@freemail.nt format, more subdomains are likely to be added eventually)
- You can login to this account on any modular device with email program
- You can send emails to other accounts, this includes replies (Copies over the reply text and other stuff, making it easier for you to maintain consistency, pretty much what an actual email would do)
- Special email address namespace exists for "service" accounts. Currently only used by "broadcast@internal-services.nt". You can't log in into these accounts, and they have special automatic behavior. The broadcaster one, for example, relays any received messages to all email accounts on the network. Has admin notifications
- Supports logging (from, to, message title) via NTNet.
- Supports AntiSpam(TM) feature, when NTNet's IDS system is enabled. Guaranteed to filter out 98% of spam, while having false positives ratio of 1% or below! Spam currently not included, but will be added as a random event.



Currently WIP. Main mail client is mostly finished, but other related things will still need some work.

atlantiscze committed:
Initial work on e-mail system
- This system is, eventually, suposed to replace PDA messages. It is slightly more complicated to use than "Click'n Send" PDAs, but compensates for that with other benefits.
- Each player is given an email account and password upon joining the game (john.smith@freemail.nt format, more subdomains are likely to be added eventually)
- You can login to this account on any modular device with email program
- You can send emails to other accounts, this includes replies (Copies over the reply text and other stuff, making it easier for you to maintain consistency, pretty much what an actual email would do)
- Special email address namespace exists for "service" accounts. Currently only used by "broadcast@internal-services.nt". You can't log in into these accounts, and they have special automatic behavior. The broadcaster one, for example, relays any received messages to all email accounts on the network. Has admin notifications
- Supports logging (from, to, message title) via NTNet.
- Supports AntiSpam(TM) feature, when NTNet's IDS system is enabled. Guaranteed to filter out 98% of spam, while having false positives ratio of 1% or below! Spam currently not included, but will be added as a random event.

Ithalan commented:
Will malicious AIs, or anyone with admin access to the network, be able to make new accounts or spoof existing ones, like it is currently possible with the message monitor console?

atlantiscze commented:
A second program (because it'd overbloat the mail client way too much) will exist, and will be usable by anyone with access_network. This program will allow for:
- Banning accounts, preventing the user from logging in.
- Creating new custom accounts.
- Reading emails in said account's inbox.
- Changing of account passwords. Due to security reasons, you will only be able to change the password, not view it.

And possibly eventually more.

WatermelonsEverywhere commented:
Will the DDOS program be able to disable this?

atlantiscze commented:
Technically, yes, as the DDoS program is capable of crashing NTNet relays, and downed NTNet relays means no NTNet connectivity to everything but consoles. It will always work on consoles though.

atlantiscze committed:
More email tweaks and changes, adds Email Administration program.

atlantiscze committed:
Final tweaks
- Hopefully should be everything for now. Adds support for attachment sending (only works on modular computers, AI view for example won't work)
- Fixes #16399

atlantiscze committed:
Merge branch 'dev' into 2017-02-20-YouveGotMail

atlantiscze commented:
This should be ready for review. Feedback/QOL suggestions are welcome too.

comma commented:
Suggestion: code that actually compiles. 
Wrong indentation at https://github.com/Baystation12/Baystation12/pull/16306/commits/4fc71c9645295d3417b5cad629aa00d97bb8a870#diff-90201b0b786589de27e0e87aea5d8bf2R15

atlantiscze committed:
Fuck that compile error
- Took three commits to fix damn indentation on one line. I must be getting old. Or git web editor sucks. Or both.

Techhead0 commented on code/game/jobs/job_controller.dm:
```diff
@@ -507,6 +507,32 @@ var/global/datum/controller/occupations/job_master
+		else
+			var/datum/computer_file/data/email_account/EA = new/datum/computer_file/data/email_account()
+			EA.password = GenerateKey()
+			EA.login = 	complete_login
```
It'd be good to pass these values to the account's `New()` proc, so you can have handling if a `login` is not supplied.

Techhead0 commented on code/modules/modular_computers/NTNet/emails/email_account.dm:
```diff
@@ -0,0 +1,82 @@
+
+/datum/computer_file/data/email_account/calculate_size()
+	size = 1
+	for(var/datum/computer_file/data/email_message/stored_message in (inbox | spam | deleted))
```
You define `all_emails()` below, might as well use it.

Techhead0 commented on maps/torch/torch_ranks.dm:
```diff
@@ -23,6 +23,7 @@
 /datum/mil_branch/expeditionary_corps
 	name = "Expeditionary Corps"
 	name_short = "SCGEC"
+	email_domain = "torch.ec.scg"
```
Not an objection, but `.gov` might be a little more fitting.

atlantiscze replied:
I have intentionally avoided using existing TLDs, to prevent accidentally using a domain that actually exists.

Techhead0 replied:
For some reason I thought .mil was an actual TLD. Carry on.

Kasuobes replied:
It can be if you make it one.

Techhead0 commented on code/modules/modular_computers/file_system/programs/research/email_administration.dm:
```diff
@@ -0,0 +1,146 @@
+	var/datum/computer_file/data/email_account/current_account = null
+	var/datum/computer_file/data/email_message/current_message = null
+	var/error = ""
+	var/list/valid_domains = list( 	"freemail.nt",\
```
Should probably be a global var, with the branch email domains adding themselves automatically. (Having hard-coded domains is okay in itself, but they shouldn't necessitate matching strings defined elsewhere.)

Techhead0 reviewed (commented):
VERY partial review. Still haven't gone through either of the two new programs, nor their nano modules.

Techhead0 commented on code/modules/modular_computers/file_system/programs/generic/email_client.dm:
```diff
@@ -0,0 +1,501 @@
+
+/datum/computer_file/program/email_client/proc/new_mail_notify()
+	for(var/mob/M in get_turf(computer))
+		to_chat(M, "\The [computer] beeps, indicating a new email has been received.")
```
This is basically a weaker visible/audible message functionality. (Which can define range, by the way)

Also, it won't notify people using consoles.

Techhead0 commented on code/modules/modular_computers/file_system/programs/generic/email_client.dm:
```diff
@@ -0,0 +1,501 @@
+	if(NME.check_message_count())
+		if(!newmail)
+			newmail = TRUE
+			new_mail_notify()
```
Doing this here means people only get notified of the first new message they receive, if they receive more than one in a row.

Techhead0 commented on code/modules/modular_computers/file_system/programs/generic/email_client.dm:
```diff
@@ -0,0 +1,501 @@
+	. = check_message_count()
+	var/list/allmails = current_account.all_emails()
+	last_message_count = allmails.len
+	update_message_count()
```
Is this meant to call itself indefinitely?

Techhead0 commented on code/modules/modular_computers/file_system/programs/generic/email_client.dm:
```diff
@@ -0,0 +1,501 @@
+
+// Persistency. Unless you log out, or unless your password changes, this will pre-fill the login data when restarting the program
+/datum/computer_file/program/email_client/kill_program()
+	newmail = FALSE
```
I notice that `kill_program` is the only place that resets `newmail`. Intentional?

Techhead0 reviewed (commented):
Few more things. Still not done.

StarSmasher44 commented:
Very cool stuff!

NullSnapshot commented:
Added Cannot merge

atlantiscze committed:
Merge branch 'dev' into 2017-02-20-YouveGotMail

atlantiscze commented:
This might fix it, assuming github's web conflict resolve isn't /that/ much dumb.

Techhead0 commented:
You replaced tabs with spaces in two spots.

Shadowlight213 commented:
:+1: This looks neat. I'll probably try porting this to tg. Might replace the PDA messenger with it in the future.

silveryferret commented:
:+1: This has given me inspiration for making 'smart phones'. I've already ported Polaris's communicators to my codebase, might see about turning communicators into a program.

comma commented:
Abandoned

atlantiscze commented:
Branch's been updated, even with few other changes and improvements i noticed when sniffing around. You can reopen this.
